analysis/topology/compact_open.lean

```diff
@@ -0,0 +1,149 @@
+/-
+Copyright (c) 2018 Reid Barton. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Reid Barton
+-/
+
+import analysis.topology.continuity tactic.tidy
+
+open set
+
+universes u v w
+
+section locally_compact
+
+-- There are various definitions of "locally compact space" in the
+-- literature, which agree for Hausdorff spaces but not in general.
+-- This one is the precise condition on X needed for the evaluation
+-- map C(X, Y) × X → Y to be continuous for all Y when C(X, Y) is
+-- given the compact-open topology.
+class locally_compact_space (α : Type u) [topological_space α] :=
+(local_compact_nhds : ∀ (x : α) (n ∈ (nhds x).sets), ∃ s ∈ (nhds x).sets, s ⊆ n ∧ compact s)
+
+variables {α : Type u} [topological_space α]
+
+lemma locally_compact_of_compact_nhds [t2_space α]
+  (h : ∀ x : α, ∃ s, s ∈ (nhds x).sets ∧ compact s) :
+  locally_compact_space α :=
+⟨assume x n hn,
+  let ⟨u, un, uo, xu⟩ := mem_nhds_sets_iff.mp hn in
+  let ⟨k, kx, kc⟩ := h x in
+  -- K is compact but not necessarily contained in N.
+  -- K \ U is again compact and doesn't contain x, so
+  -- we may find open sets V, W separating x from K \ U.
+  -- Then K \ W is a compact neighborhood of x contained in U.
+  let ⟨v, w, vo, wo, xv, kuw, vw⟩ :=
+    compact_compact_separated compact_singleton (compact_diff kc uo)
+      (by rw [singleton_inter_eq_empty]; exact λ h, h.2 xu) in
+  have wn : -w ∈ (nhds x).sets, from
+   mem_nhds_sets_iff.mpr
+     ⟨v, subset_compl_iff_disjoint.mpr vw, vo, singleton_subset_iff.mp xv⟩,
+  ⟨k - w,
+   filter.inter_mem_sets kx wn,
+   subset.trans (diff_subset_comm.mp kuw) un,
+   compact_diff kc wo⟩⟩
+
+lemma locally_compact_of_compact [t2_space α] (h : compact (univ : set α)) :
+  locally_compact_space α :=
+locally_compact_of_compact_nhds (assume x, ⟨univ, mem_nhds_sets is_open_univ trivial, h⟩)
+
+end locally_compact
+
+section compact_open
+variables {α : Type u} {β : Type v} {γ : Type w}
+variables [topological_space α] [topological_space β] [topological_space γ]
+
+local notation `C(` α `, ` β `)` := subtype (continuous : set (α → β))
+
+def compact_open_gen {s : set α} (hs : compact s) {u : set β} (hu : is_open u) : set C(α,β) :=
```

Comment:
Do we really need those `hs` and `hu`? The definition does not need them, and they will have to be repeated anyway.